### ISSUE TITLE
Type variables can be inferred as `autoType`, bypassing various type safety

### ISSUE THREAD

**Issue by mattmccutchen** [Microsoft/TypeScript#25675]
Edit: This issue affects not only evolving arrays but all control-flow evolving variables.

**TypeScript Version:**  master (60986ad)

**Search Terms:** autoType control flow evolving array push any infer return noImplicitAny

**Code** (compile with `--noImplicitAny`)

```ts
// The type of `arg` blocks inference but simplifies to T.
function logLength<T extends string, U extends string>(arg: {[K in U]: T}[U]): T {
    console.log(arg.length);
    return arg;
}
logLength(42);  // error
let z;
z = logLength(42);  // no error; T is inferred as `any`

function logFirstLength<T extends string[], U extends string>(arg: {[K in U]: T}[U]): T {
    console.log(arg[0].length);
    return arg;
}
logFirstLength([42]);  // error
let zz = [];
zz.push(logLength(42));  // no error; T is inferred as `any`
zz = logFirstLength([42]);  // no error; T is inferred as `any[]`
```

**Expected behavior:** An error on each line marked "no error".

**Actual behavior:** As marked.

**Playground Link:** [link](https://www.typescriptlang.org/play/#src=%2F%2F%20The%20type%20of%20%60arg%60%20blocks%20inference%20but%20simplifies%20to%20T.%0D%0Afunction%20logLength%3CT%20extends%20string%2C%20U%20extends%20string%3E(arg%3A%20%7B%5BK%20in%20U%5D%3A%20T%7D%5BU%5D)%3A%20T%20%7B%0D%0A%20%20%20%20console.log(arg.length)%3B%0D%0A%20%20%20%20return%20arg%3B%0D%0A%7D%0D%0AlogLength(42)%3B%20%20%2F%2F%20error%0D%0Alet%20z%3B%0D%0Az%20%3D%20logLength(42)%3B%20%20%2F%2F%20no%20error%3B%20T%20is%20inferred%20as%20%60any%60%0D%0A%0D%0Afunction%20logFirstLength%3CT%20extends%20string%5B%5D%2C%20U%20extends%20string%3E(arg%3A%20%7B%5BK%20in%20U%5D%3A%20T%7D%5BU%5D)%3A%20T%20%7B%0D%0A%20%20%20%20console.log(arg%5B0%5D.length)%3B%0D%0A%20%20%20%20return%20arg%3B%0D%0A%7D%0D%0AlogFirstLength(%5B42%5D)%3B%20%20%2F%2F%20error%0D%0Alet%20zz%20%3D%20%5B%5D%3B%0D%0Azz.push(logLength(42))%3B%20%20%2F%2F%20no%20error%3B%20T%20is%20inferred%20as%20%60any%60%0D%0Azz%20%3D%20logFirstLength(%5B42%5D)%3B%20%20%2F%2F%20no%20error%3B%20T%20is%20inferred%20as%20%60any%5B%5D%60) (remember to enable `--noImplicitAny`)

**Related Issues:** Didn't find any.

**Comment by mhegazy** [comment]
maybe we should be more conservative with `any` in return type inference.
